skimage/util/_flood_fill_cy.pyx

```diff
@@ -0,0 +1,212 @@
+    neighbor_offsets : ndarray
+        A one-dimensional array that contains the offsets to find the
+        connected neighbors for any index in `image`.
+    ravelled_seed_idx : int
+        Ravelled seed index for flood filling.
+    do_tol : unsigned char
+        If 0, only exactly equal connected points are filled and `tolerance` is
+        ignored (fastest).  Otherwise, points within `tolerance` of start value
+        will also be filled (inclusive).
+    high_tol :
+        Upper limit for tolerance comparison (ignored if `do_tol` is 0).
+    low_tol :
+        Lower limit for tolerance comparison (ignored if `do_tol` is 0).
+    """
+    cdef:
+        QueueWithHistory queue
+        unsigned char last_dim_in_neighbors
+
+    with nogil:
+        # Initialize a buffer used to queue positions while evaluating each
+        # potential maximum (flagged with 2)
+        queue_init(&queue, 64)
+        try:
+            # Conduct flood fill from this point - with or without tolerance
+            if do_tol == 0:
+                _flood_fill_do_equal(image, flags, neighbor_offsets, &queue,
+                    ravelled_seed_idx, seed_value)
+            else:
+                _flood_fill_do_tol(image, flags, neighbor_offsets, &queue,
+                    ravelled_seed_idx, seed_value, high_tol, low_tol)
+        finally:
+            # Ensure that memory is released again
+            queue_exit(&queue)
+
+
+cdef inline void _flood_fill_do_equal(
+        dtype_t[::1] image, unsigned char[::1] flags,
+        Py_ssize_t[::1] neighbor_offsets, QueueWithHistory* queue_ptr,
+        Py_ssize_t start_index, dtype_t seed_value) nogil:
+    """Fill connected areas with 1, requiring strict equality.
+
+    Parameters
+    ----------
+    image : ndarray, one-dimensional
+        The raveled view of a n-dimensional array.
+    flags : ndarray, one-dimensional
+        An array of flags that is used to store the state of each pixel during
+        evaluation.
+    neighbor_offsets : ndarray
+        A one-dimensional array that contains the offsets to find the
+        connected neighbors for any index in `image`.
+    queue_ptr :
+        Pointer to initialized queue.
+    start_index : int
+        Start position for the flood-fill.
+    seed_value :
+        Value of `image[start_index]`.
+    """
+    cdef:
+        QueueItem current_index, neighbor
+
+    # Queue start position after clearing the buffer
+    queue_clear(queue_ptr)
+    queue_push(queue_ptr, &start_index)
+    flags[start_index] = FILL
+
+    # Break loop if all queued positions were evaluated
+    while queue_pop(queue_ptr, &current_index):
+        # Look at all neighboring samples
+        for i in range(neighbor_offsets.shape[0]):
+            neighbor = current_index + neighbor_offsets[i]
+
+            # Indexing sanity check
+            if neighbor < 0:
+                continue
+            if neighbor >= image.shape[0]:
+                continue
```

Comment:
Instead of just continuing for faulty indices it may be possible to raise an error like so:
```python
if neighbor < 0 or neighbor >= image.shape[0]:
    with gil:
        raise RuntimeError("encountered invalid index {} in _flood_fill_do_equal"
                           .format(neighbor))
```
I can't image no performance penalty as acquiring the GIL is only performed if the if-statement evaluates to true but would raise a nice explicit error message.

Reply:
Before adding these there were frequent faults due to referencing invalid neighbors (easy to reproduce, just take these out and start any fill at the origin; watch the kernel crash).  These are common at edges, and just need to be excluded.

Happy to combine the if-conditionals here and elsewhere.

Reply:
Ah I see, I didn't think of the difference between how this algorithm handles the image edge vs the one in #3022.